jaraco/text/__init__.py

```diff
@@ -1,35 +1,66 @@
+from __future__ import annotations
+
 import functools
 import itertools
 import re
 import sys
 import textwrap
-from typing import Iterable
+from collections.abc import Callable, Generator, Iterable, Sequence
+from typing import (
+    TYPE_CHECKING,
+    Literal,
+    Protocol,
+    SupportsIndex,
+    TypeVar,
+    overload,
+)
+
+from jaraco.context import ExceptionTrap
+from jaraco.functools import compose, method_cache
+
+if sys.version_info >= (3, 11):
+    from importlib.resources.abc import Traversable
+else:
+    from importlib.abc import Traversable
 
 if sys.version_info >= (3, 9):
     from importlib.resources import files
 else:  # pragma: nocover
     from importlib_resources import files
 
-from jaraco.context import ExceptionTrap
-from jaraco.functools import compose, method_cache
+if TYPE_CHECKING:
+    from _typeshed import FileDescriptorOrPath, SupportsGetItem
+    from typing_extensions import Self, TypeAlias, TypeGuard, Unpack
 
+    _T_co = TypeVar("_T_co", covariant=True)
+    # Same as builtins._GetItemIterable from typeshed
+    _GetItemIterable: TypeAlias = SupportsGetItem[int, _T_co]
 
-def substitution(old, new):
+_T = TypeVar("_T")
+
+
+class _SupportsDecode(Protocol):
+    def decode(self) -> object: ...
+
+
+def substitution(old: str, new: str) -> Callable[[str], str]:
     """
     Return a function that will perform a substitution on a string
     """
     return lambda s: s.replace(old, new)
 
 
-def multi_substitution(*substitutions):
+def multi_substitution(*_substitutions: str) -> Callable[[str], str]:
```

Comment:
This can be avoided in mypy by using `allow_redefinition = True` in configs. But then will fail pyright: (tbf pyright/pylance is also not directly checked in this project)
```suggestion
def multi_substitution(*substitutions: str) -> Callable[[str], str]:  # pyright: ignore[reportRedeclaration]
```

Also alternatively:
```py
def multi_substitution(*substitutions: str) -> Callable[[str], str]:
    """
    Take a sequence of pairs specifying substitutions, and create
    a function that performs those substitutions.

    >>> multi_substitution(('foo', 'bar'), ('bar', 'baz'))('foo')
    'baz'
    """
    # compose function applies last function first, so reverse the
    #  substitutions to get the expected order.
    return compose(*reversed(tuple(itertools.starmap(substitution, substitutions))))
```